csrc/scheduler/mma_utils.cpp

```diff
@@ -1229,6 +1229,158 @@ RolesMapOpt getTensorsRoles(Fusion* fusion) {
+  }
+
+  // Also ensure that it's not the same dim for the two muls. that's
+  // getting broadcast.
+  auto idx_l = std::find(bcastFlags_l.begin(), bcastFlags_l.end(), true) -
+      bcastFlags_l.begin();
+  auto idx_r = std::find(bcastFlags_r.begin(), bcastFlags_r.end(), true) -
+      bcastFlags_r.begin();
+  if (idx_l == idx_r) {
+    return false;
+  }
+
+  // Also ensure that the reduction dim is not either of the broadcast dim.
+  if (reduction_axis == idx_l || reduction_axis == idx_r) {
+    return false;
+  }
+
+  // Check different dimensions are the broadcast dims.
+  return true;
+}
+
+TensorView* getTensorviewPriorToCast(TensorView* in) {
+  if (in->getDataType() == DataType::BFloat16 ||
+      in->getDataType() == DataType::Half) {
+    return in;
+  }
+  TensorView* ret = nullptr;
+  if (in->getDataType() == DataType::Float) {
+    // get the producer.
+    auto* cast = in->definition();
+    auto* uCastOp = dynamic_cast<UnaryOp*>(cast);
+    if (uCastOp && uCastOp->getUnaryOpType() == UnaryOpType::Cast) {
+      auto* aTV = static_cast<TensorView*>(uCastOp->in());
+      NVF_CHECK(
+          (aTV->getDataType() == DataType::BFloat16 ||
+           aTV->getDataType() == DataType::Half),
+          "The output of the cast must be half or BF16");
+      ret = aTV;
+    }
+  }
+  return ret;
+}
+
+std::optional<MulSumAsMmaProps> getMulSumInsOutsBcasts(
+    BinaryOp* mop,
+    ReductionOp* redop) {
+  auto a = getTensorviewPriorToCast(static_cast<TensorView*>(mop->lhs()));
+  auto b = getTensorviewPriorToCast(static_cast<TensorView*>(mop->rhs()));
+
+  // Get the dimension of the reduction in the output. If not present, bail.
+  auto red_axis =
+      static_cast<TensorView*>(redop->out())->domain()->getReductionAxis();
+  if (!red_axis.has_value()) {
+    return std::nullopt;
+  }
+
+  if (broadcastsAreValid(a, b, *red_axis)) {
+    return MulSumAsMmaProps(
+        mop,
+        redop,
+        a,
+        b,
+        static_cast<TensorView*>(redop->output(0)),
+        dynamic_cast<BroadcastOp*>(a->definition()),
+        dynamic_cast<BroadcastOp*>(b->definition()));
+  }
+  return std::nullopt;
+}
+} // namespace
+
+void CombineMulSum::handle(ReductionOp* stmt) {
+  // Check if operation is a sum.
+  if (stmt->getReductionOpType() == BinaryOpType::Add) {
+    auto* inputOfSum = stmt->in();
+    if (inputOfSum != nullptr) {
+      auto* expr = inputOfSum->definition();
+      // Then check if the prodcer of the sum is a mul.
+      if (auto bOp = dynamic_cast<BinaryOp*>(expr)) {
+        // If it'a mul followed by a sum, put this in a list.
+        if (bOp->getBinaryOpType() == BinaryOpType::Mul) {
+          auto props = getMulSumInsOutsBcasts(bOp, stmt);
+          if (props.has_value()) {
+            mul_sum_props_.push_back(*props);
+          }
+        }
+      }
+    }
+  }
+};
+
+std::vector<MulSumAsMmaProps> CombineMulSum::generateMulSumCanidates() {
+  if (!mul_sum_props_.empty()) {
+    return mul_sum_props_;
+  }
```

Comment:
Does this mean this function is caching its result? If so, please indicate that in the doxygen comment.

Reply:
Also, if this is needed to prevent repeated traversals, keep in mind that in the vast majority of cases this vector will always be empty, so you might want to use a flag to explicitly indicate whether it has been processed. The above-mentioned comment should make it more clear how likely we are to be making redundant calls to this method.